Automattic/wp-calypso
# Extensions: Add extension redux state

***Notice***
This PR is being deferred in favor of a more static approach for now (#11728). However, the dynamic approach will need to be revisited at some point in the future. Issue #12582 was created to track the future implementation of this enhancement.

Note: I'd like to keep this PR open until it is implemented, so as to not lose the code that exists on this branch until it has been successfully re-implemented.
***


This PR adds redux state support to extensions by way of enabling dynamic addition of reducers to the store.

It also creates a `withReducer()` Higher Order Component which ties a reducer to a React component for its lifecycle (i.e. reducer is added upon mount, removed upon dismount) This allows the reducer to be present for the duration of the use of the extension, but not be present when the extension is not in use.

Note: This triggers a console error log from redux's
combineReducers() function. This is caused by the persistence
of top-level state keys brought in to the initial state before
the extension has added its reducer. It's not a problem, but
should become a discussion point for what's next. (Issue logged as #11455)

To Test:
1. Run `make run`
2. Visit `http://calypso.localhost:3000/hello-dolly`
3. Verify that the lyrics show up in bold.
4. Click 'Next' to see the next lyric.

Still to be done:

* [x] Add reducer support.
* [x] Add Hello Dolly reducer.
* [x] Add actions to Hello Dolly example.
* [x] Add Lyric UI component to Hello Dolly example which reads from Redux state.

### Pre-Merge Checklist ###

 - [ ] Tests in Docker
 - [ ] Clear the browser caches `localStorage.clear(); indexedDB.deleteDatabase( 'calypso' )`
 - [ ] Sign in outside of the proxy and create a user and a site
 - [ ] Write and publish a post
 - [ ] Test on staging after merge and before deploy
